packages/console/src/contexts/TenantsProvider.tsx

```diff
@@ -1,79 +1,127 @@
 import { defaultManagementApi } from '@logto/schemas';
 import { type TenantInfo, TenantTag } from '@logto/schemas/models';
-import { conditional, noop } from '@silverhand/essentials';
+import { conditionalArray, noop } from '@silverhand/essentials';
 import type { ReactNode } from 'react';
 import { useCallback, useMemo, createContext, useState } from 'react';
 import type { NavigateOptions } from 'react-router-dom';
 
 import { isCloud } from '@/consts/env';
 import { getUserTenantId } from '@/consts/tenants';
 
-type Props = {
-  children: ReactNode;
-};
+type CurrentTenantStatus = 'pending' | 'validating' | 'validated';
 
+/** @see {@link TenantsProvider} for why `useSWR()` is not applicable for this context. */
 type Tenants = {
-  tenants?: TenantInfo[];
-  isSettle: boolean;
-  setTenants: (tenants: TenantInfo[]) => void;
-  setIsSettle: (isSettle: boolean) => void;
+  tenants: readonly TenantInfo[];
+  /** Indicates if the tenants data is ready for the first render. */
+  isInitComplete: boolean;
+  /** Reset tenants to the given value. It will overwrite the current tenants data and set `isInitComplete` to `true`. */
+  resetTenants: (tenants: TenantInfo[]) => void;
```

Comment:
How about providing a `reloadTenants: () => void` function instead of `resetTenants: (tenants: TenantInfo[]) => void`? Then we can encapsulate `cloudApi get tenants` in the context provider maybe?

Reply:
i was thinking about this. but currently our none of api calls seats in the data store, instead they spread across components.

so if we are moving those async api calls it's more like a pattern shift? idk

Reply:
Hmmm, OK. You are making a point here, I guess. Let's move on